File: obsidian-mcp-ts/src/obsidian/folders.ts
import fs from "fs";
import path from "path";
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { getVaultPath } from "./vault.js";
import { findMarkdownFiles, resolveNotePath, relPath, stem } from "./utils.js";

function findAllDirs(dir: string, vaultPath: string): string[] {
  const results: string[] = [];
  function walk(current: string) {
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(current, { withFileTypes: true });
    } catch { return; }
    for (const entry of entries) {
      if (!entry.isDirectory() || entry.name === ".obsidian") continue;
      const full = path.join(current, entry.name);
      results.push(path.relative(vaultPath, full));
      walk(full);
    }
  }
  walk(dir);
  return results;
}

export function registerFolderTools(server: McpServer) {
  server.tool(
    "create_folder",
    "Create a new folder in the vault",
    { folder_path: z.string().describe("Path to the folder (e.g., 'Projects/Tasks')") },
    async ({ folder_path }) => {
      try {
        const vaultPath = getVaultPath();
        const newFolder = path.join(vaultPath, folder_path);

        if (fs.existsSync(newFolder)) {
          return { content: [{ type: "text" as const, text: JSON.stringify({ success: true, message: `Folder already exists: ${folder_path}`, folder_path }) }] };
        }

        fs.mkdirSync(newFolder, { recursive: true });
        return { content: [{ type: "text" as const, text: JSON.stringify({ success: true, folder_path, full_path: newFolder }) }] };
      } catch (e: any) {
        return { content: [{ type: "text" as const, text: JSON.stringify({ error: e.message }) }] };
      }
    }
  );

  server.tool(
    "move_note",
    "Move a note to a different folder",
    {
      note_name: z.string().describe("Name of the note (without .md extension) or current relative path"),
      new_folder: z.string().describe("Destination folder path"),
    },
    async ({ note_name, new_folder }) => {
      try {
        const vaultPath = getVaultPath();
        const notePath = resolveNotePath(vaultPath, note_name);
        if (!notePath) return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Note not found: ${note_name}` }) }] };

        const destFolder = path.join(vaultPath, new_folder);
        fs.mkdirSync(destFolder, { recursive: true });

        const destPath = path.join(destFolder, path.basename(notePath));
        if (fs.existsSync(destPath)) {
          return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Note already exists in destination: ${new_folder}/${path.basename(notePath)}` }) }] };
        }

        fs.renameSync(notePath, destPath);
        return {
          content: [{
            type: "text" as const,
            text: JSON.stringify({ success: true, note_name: stem(notePath), old_path: relPath(vaultPath, notePath), new_path: relPath(vaultPath, destPath), new_folder }),
          }],
        };
      } catch (e: any) {
        return { content: [{ type: "text" as const, text: JSON.stringify({ error: e.message }) }] };
      }
    }
  );

  server.tool(
    "list_folders",
    "List all folders in the vault",
    { folder: z.string().optional().describe("Optional parent folder to list subfolders from") },
    async ({ folder }) => {
      try {
        const vaultPath = getVaultPath();
        const searchPath = folder ? path.join(vaultPath, folder) : vaultPath;
        if (!fs.existsSync(searchPath)) return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Folder not found: ${folder}`, folders: [] }) }] };

        const folders = findAllDirs(searchPath, vaultPath);
        if (!folder) folders.unshift(".");
        folders.sort();

        return { content: [{ type: "text" as const, text: JSON.stringify({ parent_folder: folder || "root", total_folders: folders.length, folders }) }] };
      } catch (e: any) {
        return { content: [{ type: "text" as const, text: JSON.stringify({ error: e.message, folders: [] }) }] };
      }
    }
  );

  server.tool(
    "get_folder_statistics",
    "Get statistics about a specific folder",
    { folder: z.string().describe("Folder path (e.g., 'Projects')") },
    async ({ folder }) => {
      try {
        const vaultPath = getVaultPath();
        const folderPath = path.join(vaultPath, folder);

        if (!fs.existsSync(folderPath)) return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Folder not found: ${folder}` }) }] };
        if (!fs.statSync(folderPath).isDirectory()) return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Path is not a folder: ${folder}` }) }] };

        let totalNotes = 0, totalSize = 0, totalLines = 0;
        const subfolders: string[] = [];
        const noteList: any[] = [];

        for (const entry of fs.readdirSync(folderPath, { withFileTypes: true })) {
          const fullPath = path.join(folderPath, entry.name);
          if (entry.isDirectory() && entry.name !== ".obsidian") {
            subfolders.push(entry.name);
          } else if (entry.isFile() && entry.name.endsWith(".md")) {
            try {
              const content = fs.readFileSync(fullPath, "utf-8");
              totalNotes++;
              totalSize += content.length;
              const lines = content.split("\n").length;
              totalLines += lines;
              noteList.push({ name: path.basename(entry.name, ".md"), size: content.length, lines });
            } catch { continue; }
          }
        }

        return {
          content: [{
            type: "text" as const,
            text: JSON.stringify({
              folder,
              total_notes: totalNotes,
              total_subfolders: subfolders.length,
              subfolders: subfolders.sort(),
              total_size_bytes: totalSize,
              total_size_kb: Math.round((totalSize / 1024) * 100) / 100,
              total_lines: totalLines,
              average_note_size: totalNotes ? Math.round((totalSize / totalNotes) * 100) / 100 : 0,
              average_lines_per_note: totalNotes ? Math.round((totalLines / totalNotes) * 100) / 100 : 0,
              notes: noteList.sort((a, b) => a.name.localeCompare(b.name)),
            }),
          }],
        };
      } catch (e: any) {
        return { content: [{ type: "text" as const, text: JSON.stringify({ error: e.message }) }] };
      }
    }
  );

  server.tool(
    "delete_folder",
    "Delete a folder from the vault",
    {
      folder_path: z.string().describe("Path to the folder to delete"),
      force: z.boolean().default(false).describe("If true, delete folder even if it contains notes"),
    },
    async ({ folder_path, force }) => {
      try {
        const vaultPath = getVaultPath();
        const folderFull = path.join(vaultPath, folder_path);

        if (!fs.existsSync(folderFull)) return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Folder not found: ${folder_path}` }) }] };
        if (!fs.statSync(folderFull).isDirectory()) return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Path is not a folder: ${folder_path}` }) }] };

        const noteCount = findMarkdownFiles(folderFull).length;
        if (noteCount > 0 && !force) {
          return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Folder contains ${noteCount} note(s). Use force=true to delete anyway.`, note_count: noteCount }) }] };
        }

        fs.rmSync(folderFull, { recursive: true, force: true });
        return { content: [{ type: "text" as const, text: JSON.stringify({ success: true, folder_path, notes_deleted: noteCount }) }] };
      } catch (e: any) {
        return { content: [{ type: "text" as const, text: JSON.stringify({ error: e.message }) }] };
      }
    }
  );
}

File: obsidian-mcp-py/obsidian/folders.py
"""
Obsidian Folders Management

Provides tools for managing folders in the vault.
"""

from pathlib import Path
from typing import Optional, List
from obsidian.vault import get_vault_path, safe_path


def register_folder_tools(mcp):
    """
    Register all folder-related MCP tools.
    
    Args:
        mcp: FastMCP instance to register tools with
    """
    
    @mcp.tool()
    def create_folder(folder_path: str) -> dict:
        """
        Create a new folder in the vault.
        
        Args:
            folder_path: Path to the folder (e.g., "Projects/Tasks")
        
        Returns:
            Dictionary with creation status
        """
        try:
            vault_path = get_vault_path()
            new_folder = safe_path(vault_path, folder_path)
            
            if new_folder.exists():
                return {
                    "success": True,
                    "message": f"Folder already exists: {folder_path}",
                    "folder_path": folder_path
                }
            
            new_folder.mkdir(parents=True, exist_ok=True)
            
            return {
                "success": True,
                "folder_path": folder_path,
                "full_path": str(new_folder)
            }
        except Exception as e:
            return {"error": str(e)}
    
    
    @mcp.tool()
    def move_note(note_name: str, new_folder: str) -> dict:
        """
        Move a note to a different folder.
        
        Args:
            note_name: Name of the note (without .md extension) or current relative path
            new_folder: Destination folder path (e.g., "Projects/Tasks")
        
        Returns:
            Dictionary with move status
        """
        try:
            vault_path = get_vault_path()
            
            # Find source note
            note_path = safe_path(vault_path, f"{note_name}.md")
            if not note_path.exists():
                note_path = safe_path(vault_path, note_name)
            
            if not note_path.exists():
                return {"error": f"Note not found: {note_name}"}
            
            # Create destination folder if needed
            dest_folder = safe_path(vault_path, new_folder)
            dest_folder.mkdir(parents=True, exist_ok=True)
            
            # Move note
            dest_path = dest_folder / note_path.name
            if dest_path.exists():
                return {"error": f"Note already exists in destination: {new_folder}/{note_path.name}"}
            
            note_path.rename(dest_path)
            
            return {
                "success": True,
                "note_name": note_path.stem,
                "old_path": str(note_path.relative_to(vault_path)),
                "new_path": str(dest_path.relative_to(vault_path)),
                "new_folder": new_folder
            }
        except Exception as e:
            return {"error": str(e)}
    
    
    @mcp.tool()
    def list_folders(folder: Optional[str] = None) -> dict:
        """
        List all folders in the vault.
        
        Args:
            folder: Optional parent folder to list subfolders from
        
        Returns:
            Dictionary with list of folders
        """
        try:
            vault_path = get_vault_path()
            
            if folder:
                search_path = safe_path(vault_path, folder)
            else:
                search_path = vault_path
            
            if not search_path.exists():
                return {"error": f"Folder not found: {folder}", "folders": []}
            
            folders = set()
            
            # Find all directories
            for item in search_path.rglob("*"):
                if item.is_dir() and ".obsidian" not in item.parts:
                    rel_path = item.relative_to(vault_path)
                    folders.add(str(rel_path))
            
            # Also include the root
            if not folder:
                folders.add(".")
            
            folders_list = sorted(list(folders))
            
            return {
                "parent_folder": folder or "root",
                "total_folders": len(folders_list),
                "folders": folders_list
            }
        except Exception as e:
            return {"error": str(e), "folders": []}
    
    
    @mcp.tool()
    def get_folder_statistics(folder: str) -> dict:
        """
        Get statistics about a specific folder.
        
        Args:
            folder: Folder path (e.g., "Projects")
        
        Returns:
            Dictionary with folder statistics
        """
        try:
            vault_path = get_vault_path()
            folder_path = safe_path(vault_path, folder)
            
            if not folder_path.exists():
                return {"error": f"Folder not found: {folder}"}
            
            if not folder_path.is_dir():
                return {"error": f"Path is not a folder: {folder}"}
            
            total_notes = 0
            total_size = 0
            total_lines = 0
            subfolders = []
            note_list = []
            
            for item in folder_path.iterdir():
                if item.is_dir():
                    if ".obsidian" not in item.parts:
                        subfolders.append(item.name)
                elif item.suffix == ".md":
                    try:
                        content = item.read_text(encoding="utf-8")
                        total_notes += 1
                        total_size += len(content)
                        total_lines += len(content.splitlines())
                        note_list.append({
                            "name": item.stem,
                            "size": len(content),
                            "lines": len(content.splitlines())
                        })
                    except Exception:
                        continue
            
            # Calculate averages
            avg_size = total_size / total_notes if total_notes > 0 else 0
            avg_lines = total_lines / total_notes if total_notes > 0 else 0
            
            return {
                "folder": folder,
                "total_notes": total_notes,
                "total_subfolders": len(subfolders),
                "subfolders": sorted(subfolders),
                "total_size_bytes": total_size,
                "total_size_kb": round(total_size / 1024, 2),
                "total_lines": total_lines,
                "average_note_size": round(avg_size, 2),
                "average_lines_per_note": round(avg_lines, 2),
                "notes": sorted(note_list, key=lambda x: x["name"])
            }
        except Exception as e:
            return {"error": str(e)}
    
    
    @mcp.tool()
    def delete_folder(folder_path: str, force: bool = False) -> dict:
        """
        Delete a folder from the vault.
        
        Args:
            folder_path: Path to the folder to delete
            force: If True, delete folder even if it contains notes
        
        Returns:
            Dictionary with deletion status
        """
        try:
            vault_path = get_vault_path()
            folder = safe_path(vault_path, folder_path)
            
            if not folder.exists():
                return {"error": f"Folder not found: {folder_path}"}
            
            if not folder.is_dir():
                return {"error": f"Path is not a folder: {folder_path}"}
            
            # Count notes in folder
            note_count = len(list(folder.rglob("*.md")))
            
            if note_count > 0 and not force:
                return {
                    "error": f"Folder contains {note_count} note(s). Use force=True to delete anyway.",
                    "note_count": note_count
                }
            
            # Delete folder
            import shutil
            shutil.rmtree(folder)
            
            return {
                "success": True,
                "folder_path": folder_path,
                "notes_deleted": note_count
            }
        except Exception as e:
            return {"error": str(e)}
